crates/autopilot/src/database/onchain_order_events.rs

```diff
@@ -56,6 +58,10 @@ pub struct OnchainOrderParser<EventData: Send + Sync, EventRow: Send + Sync> {
     settlement_contract: H160,
     liquidity_order_owners: HashSet<H160>,
     metrics: &'static Metrics,
+    /// A helper to keep track of metrics changes on reorg. It collects orders
+    /// that were already considered in the metrics so that they can be deleted
+    /// in case a reorg occurs.
+    revertible_orders: OrdersByBlockNumber,
 }
 
 impl<EventData, EventRow> OnchainOrderParser<EventData, EventRow>
```

Comment:
The design of `OnchainOrderParser` is weird.

There's an `OnchainOrderParsing` trait which is not implemented for `OnchainOrderParser`, but the `OnchainOrderParser` does wrap a `Box<dyn OnchainOrderParsing>`. The comment for `OnchainOrderParsing` (which is not a doc comment) reads:

```
// The following trait allows to implement custom onchain order parsing for
// differently placed orders. E.g., there will be a implementation for ethflow
// and presign orders. For each of the customs types, the trait allows to
// implement parsing the on-chain data and storing the event data
```

However the trait is in fact only implemented for your `EthFlowOnchainOrderParser` type and that is the single implementation of it. :question: 

It's also odd that this module is a submodule of `database` when it seems to be doing _a lot_ more than database access.

Reply:
>However the trait is in fact only implemented for your EthFlowOnchainOrderParser type and that is the single implementation of it. ❓

We created that general concept on the [OnchainOrder defined via this interface ](https://github.com/cowprotocol/ethflowcontract/blob/main/src/interfaces/ICoWSwapOnchainOrders.sol) that can be inherited by any contract. Yes, currently there is only one implementation, but we intended to add other ones for each contract.

>It's also odd that this module is a submodule of database when it seems to be doing a lot more than database access.

yeah, you are right about this aspect. We should give it its own module.

Reply:
The idea is that all ochain orders are created by emitting an EVM event of a standard format ([this](https://github.com/cowprotocol/ethflowcontract/blob/9c74c8ba36ff9ff3e255172b02454f831c066865/src/interfaces/ICoWSwapOnchainOrders.sol#L39-L44)). This is not specific to eth-flow orders, but indeed eth-flow orders are, right now, the only contract that makes use of that.

When reading an order from the blockchain, however, there might be the need for reading data that is specific to a particular implementation (from [this field](https://github.com/cowprotocol/ethflowcontract/blob/9c74c8ba36ff9ff3e255172b02454f831c066865/src/interfaces/ICoWSwapOnchainOrders.sol#L43)). How the data is parsed changes based on what implements onchain orders, and what we expect from this parsing is specified by `OnchainOrderParsing`. This is also why only  `EthFlowOnchainOrderParser` implements it, since only eth-flow orders are creating onchain orders right now.

I agree the naming is confusing. Maybe `OnchainOrderParsing` should be `OnchainOrderDataParsing`?